standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -604,15 +630,35 @@ public OpenTxnsResponse openTxns(OpenTxnRequest rqst) throws MetaException {
   @Override
   @RetrySemantics.Idempotent
   public void abortTxn(AbortTxnRequest rqst) throws NoSuchTxnException, MetaException, TxnAbortedException {
-    long txnid = rqst.getTxnid();
+    long txnid;
+    long sourceTxnId = -1;
+    ResultSet rs;
     try {
       Connection dbConn = null;
       Statement stmt = null;
       try {
         lockInternal();
         dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
+        stmt = dbConn.createStatement();
+
+        if (rqst.isSetReplPolicy()) {
+          sourceTxnId = rqst.getTxnid();
+          String query = "select TM_TARGET_TXN_ID from REPL_TXN_MAP where TM_SRC_TXN_ID = " + sourceTxnId +
+              " and TM_REPL_POLICY = " + quoteString(rqst.getReplPolicy());
+          String s = sqlGenerator.addForUpdateClause(query);
+          LOG.debug("Going to execute query <" + s + ">");
+          rs = stmt.executeQuery(s);
+          if (!rs.next()) {
+            LOG.debug("Target txn for Transaction not present <" +
+                    quoteString(rqst.getReplPolicy()) +":" +sourceTxnId + ">");
+            return;
+          }
+          txnid = rs.getLong(1);
+        } else {
+          txnid = rqst.getTxnid();
+        }
+
         if (abortTxns(dbConn, Collections.singletonList(txnid), true) != 1) {
-          stmt = dbConn.createStatement();
           TxnStatus status = findTxnState(txnid,stmt);
           if(status == TxnStatus.ABORTED) {
             LOG.info("abortTxn(" + JavaUtils.txnIdToString(txnid) +
```

Comment:
If the txn is already aborted/committed, then also need to remove the entry from REPL_TXN_MAP.
Same for commitTxn as well.

Reply:
i think in case of replPolicy is set the first check for target txn id will only return ..so for replay case this scenario will never hit.